pandas/io/parsers.py

```diff
@@ -1425,12 +1425,14 @@ def __init__(self, src, **kwds):
         self.orig_names = self.names[:]
 
         if self.usecols:
-            if len(self.names) > len(self.usecols):
-                self.names = [n for i, n in enumerate(self.names)
-                              if (i in self.usecols or n in self.usecols)]
-
-            if len(self.names) < len(self.usecols):
-                raise ValueError("Usecols do not match names.")
+            if self._reader.file_header is not None:
+                h = self._reader.file_header[0]
+                usecol_len = len(set(self.usecols) - set(h))
+                usecoli_len = len(set(self.usecols) - set(range(0, len(h))))
+                if usecol_len > 0 and usecoli_len > 0:
+                    raise ValueError("Usecols do not match names.")
```

Comment:
From what I can tell the way the C parser works is that it will populate the names from the names value passed into the TextReader class. So if I wanted to read columns 'A' and 'C' from a CSV but rename them as 'Category' and 'Total' (names parameter) usecols compares against the names parameter

Another example (which happens with the pandas.io.tests.parser.test_parsers.TestCParserLowMemory test):
        s = """0,1,20140101,0900,4
        0,1,20140102,1000,4"""
        parse_dates = [[1, 2]]
        names = list('acd')

        df = self.read_csv(StringIO(s), names=names,
                           usecols=[0, 2, 3],
                           parse_dates=parse_dates)
Fails because index 3 doesn't exist in the names list passed in (it's only 0-2)